Provide NavigationEventDispatcherOwner for non-android targets

terrakok opened the pull request:
Suppot the `NavigationEventDispatcherOwner` for a correct Navigation3 support on non-android targets.

Fixes [CMP-7646](https://youtrack.jetbrains.com/issue/CMP-7646) Support Navigation 3 for Compose Multiplatform

## Testing
Manually checked the demo app

## Release Notes
### Highlights - Navigation
- Provide required `NavigationEventDispatcherOwner` for a correct Navigation3 support.

terrakok committed:
Add navigation event handling support for Desktop in Compose UI

terrakok committed:
Use default NavigationEventDispatcherOwner for non-Android targets in Compose.

terrakok committed:
Replace BackGestureDispatcher with NavigationEventInput for iOS Integration

terrakok committed:
Remove DesktopBackGestureDispatcher in favor of NavigationEventInput integration.

terrakok committed:
Remove unused `ui-backhandler` module dependency from Compose UI

terrakok committed:
Remove experimental `enableBackGesture` flag from `ComposeUIViewControllerConfiguration`

terrakok committed:
Migrate the ui-backhandler module on the navigationevent library

terrakok committed:
Extract `navigationevent` version to a project property for better configurability.

terrakok committed:
Dump API

terrakok committed:
Add null-check for `LocalNavigationEventDispatcherOwner` in `BackHandler`

MatkovIvan commented on MULTIPLATFORM.md:
```diff
@@ -90,13 +90,14 @@ Compose Multiplatform core libraries can be published to local Maven with the fo
    `-Pjetbrains.publication.version.LIFECYCLE`,
    `-Pjetbrains.publication.version.NAVIGATION`,
    `-Pjetbrains.publication.version.NAVIGATION_3`,
+   `-Pjetbrains.publication.version.NAVIGATION_EVENT`,
```
Could you please rebase to remove already merged diff?

MatkovIvan commented on compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/navigationevent/UIKitNavigationEventDispatcherOwner.kt:
```diff
@@ -0,0 +1,24 @@
+import androidx.navigationevent.NavigationEventDispatcher
+import androidx.navigationevent.NavigationEventDispatcherOwner
+
+internal class UIKitNavigationEventDispatcherOwner : NavigationEventDispatcherOwner {
```
It seems it should be combined with other owners (it's not only about "lifecycle" already)

terrakok replied:
I'd let it as a refactoring for the ios team. it is not related with the feature.

ASalavei replied:
> It seems it should be combined with other owners

Unlike other owners, UIKitNavigationEventDispatcherOwner is used separately for the root scene and layer scenes. No need to combine it with IOSLifecycleOwner.

MatkovIvan replied:
> it is not related with the feature

It does. It doesn't follow the existing pattern

> Unlike other owners

I see no difference with `ViewModelStoreOwner`

ASalavei replied:
> I see no difference with ViewModelStoreOwner

Because we're using 2 sets of gesture handlers, i'd like to see them controlled individually - it cannot be done currently with a single NavigationEventDispatcherOwner

terrakok replied:
owner can be single. inputs should be different

terrakok replied:
I'll refactor this soon

terrakok replied:
done

MatkovIvan commented on compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/uikit/ComposeUIViewControllerConfiguration.uikit.kt:
```diff
@@ -67,12 +67,6 @@ class ComposeUIViewControllerConfiguration {
-     * A flag to enable or disable iOS BackGesture recognizer.
-     */
-    @ExperimentalComposeUiApi
-    var enableBackGesture: Boolean = true
```
Is it OK to remove this ability? cc @ASalavei

terrakok replied:
yes. It was an experimental flag. I discussed it with google. now, there is an official way to disable it:
```
val navEventDispatcherOwner = LocalNavigationEventDispatcherOwner.current
LaunchedEffect(Unit) {
  navEventDispatcherOwner?.navigationEventDispatcher?.isEnabled = false
}
```

ASalavei commented on compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/navigationevent/UIKitNavigationEventInput.kt:
```diff
@@ -0,0 +1,262 @@
+    }
+
+    private var dispatcher: NavigationEventDispatcher? = null
+    private val isEnabled get() = dispatcher?.isEnabled == true
```
For me, `NavigationEventDispatcher` is always enabled - no matter if there are active listeners in Compose scene or not.
From iOS side, `UIKitBackGestureRecognizer` should be enabled only on demand, not always.

terrakok replied:
you are right. thank you 👍

terrakok committed:
Enable ios gesture recognizers only when there are active callbacks.

terrakok committed:
Introduce `UIKitArchitectureComponentsOwner` with Lifecycle, ViewModelStore and NavigationEventDispatcher owners.

And instead of a new NavigationEventDispatcher in the UIKitComposeSceneLayer, a new UIKitNavigationEventInput is used

terrakok committed:
Add Popup with Predictive Back Handler in the iOS demo.

terrakok commented:
checked iOS layers with a dialog and a popup. works like a charm

https://github.com/user-attachments/assets/b5c81d24-0f04-461b-9e70-d0810eeeda6e

terrakok committed:
Remove unused `UIKitNavigationEventDispatcherOwner` and related imports.

terrakok committed:
Move `navigationevent-compose` dependency in `ui-backhandler` module to the skiko only source set.

terrakok committed:
Deprecate `BackEventCompat` and `BackHandler` APIs in favor of `NavigationEvent` and `NavigationEventHandler`. Remove unused dependencies and dump API.

MatkovIvan commented on compose/ui/ui-backhandler/src/androidMain/kotlin/androidx/compose/ui/backhandler/BackHandler.android.kt:
```diff
@@ -17,17 +17,16 @@
-import androidx.compose.ui.ExperimentalComposeUiApi
 import kotlinx.coroutines.flow.Flow
 
-@ExperimentalComposeUiApi
```
We must keep it as `ExperimentalComposeUiApi` even after deprecation. It should be an addition, not replacement

MatkovIvan reviewed (changes requested):
Revert experimental annotations. Removing it means stabilization that we do not want to

terrakok committed:
Return `@ExperimentalComposeUiApi` in `ui-backhandler` module.

MatkovIvan commented on compose/ui/ui-backhandler/src/androidMain/kotlin/androidx/compose/ui/backhandler/BackHandler.android.kt:
```diff
@@ -20,13 +20,15 @@ import androidx.compose.runtime.Composable
 import androidx.compose.ui.ExperimentalComposeUiApi
 import kotlinx.coroutines.flow.Flow
 
+@Deprecated("Use NavigationEventHandler instead")
```
```suggestion
@Deprecated(
    message = "Use a {@link androidx.navigationevent.compose.NavigationEventHandler} instead",
    replaceWith = ReplaceWith(
        expression = "NavigationEventHandler(enabled, onBack)",
        imports = arrayOf("androidx.navigationevent.compose.NavigationEventHandler"),
    ),
    level = DeprecationLevel.WARNING, // TODO: Replace to ERROR in 1.11
)
```

terrakok replied:
[KTIJ-35595](https://youtrack.jetbrains.com/issue/KTIJ-35595) ReplaceWith works wrong with expect/actual and default arguments

MatkovIvan reviewed (commented):
LGTM in general, but let's add proper metadata in depracation annotations to help with migration

ASalavei commented on compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/platform/UIKitArchitectureComponentsOwner.kt:
```diff
@@ -22,10 +22,16 @@ import androidx.lifecycle.LifecycleRegistry
+import androidx.navigationevent.NavigationEventDispatcherOwner
 
-internal class IOSLifecycleOwner: LifecycleOwner, ViewModelStoreOwner {
+internal class UIKitArchitectureComponentsOwner :
```
I'm not happy with the idea to put everything in a single class by the criteria that they are Owners interfaces.
Also, ViewModelStore placed here incorrectly. ViewModelStoreOwner and NavigationEventDispatcher should be implemented by the separate instances - either that can be separate classes or anonymous object.
It's up to you - you can do it in the current MR, or I'll fix it by myself later.

terrakok replied:
Since I've already refactored this part of the code let's give it to you. ok? I have no strict preferences here but it seems logical though because on other platforms we have these interfaces in a single place

ASalavei commented on compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/scene/UIKitComposeSceneLayer.uikit.kt:
```diff
@@ -180,11 +179,21 @@ internal class UIKitComposeSceneLayer(
-        content = content
-    )
+    ) {
+        val navigationEventDispatcherOwner = LocalInternalNavigationEventDispatcherOwner.current
```
Two different approaches where used in `ComposeHostingViewController` and here. Either this code should be used in ComposeSceneMediator once or it's looks like the optimal solution will be to:
- make the `navigationEventInput` internal
- Add input at the place where `UIKitComposeSceneLayer` is created
- use `onClose` callback to remove input

ASalavei commented on compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/scene/ComposeHostingViewController.uikit.kt:
```diff
@@ -122,12 +122,10 @@ internal class ComposeHostingViewController(
         density = rootView.density,
         getTopLeftOffsetInWindow = { IntOffset.Zero } //full screen
-    )
+    ).also { eventInput -> archComponentsOwner.navigationEventDispatcher.addInput(eventInput) }
```
The `eventInput` added here never removes. It should not be a big issue since `navigationEventDispatcher` and `eventInput` have the same lifetime, but it would be better to add event input in the `viewControllerDidEnterWindowHierarchy` and remove in `viewControllerDidLeaveWindowHierarchy`

terrakok committed:
Integrate `NavigationEventDispatcher` into `UIKitComposeSceneLayer`.

igordmn commented on compose/ui/ui/build.gradle:
```diff
@@ -206,7 +206,8 @@ if(AndroidXComposePlugin.isMultiplatformEnabled(project)) {
                     api(project(":compose:ui:ui-text"))
                     api(libs.skikoCommon)
                     implementation(libs.atomicFu)
-                    implementation(project(":compose:ui:ui-backhandler"))
```
This probably broke backward compatibility for klibs due to [KT-60874](https://youtrack.jetbrains.com/issue/KT-60874) [KLIB Resolve] Don't save transitive dependencies in manifest's `depends` property.

Please [fix it](https://youtrack.jetbrains.com/issue/CMP-9008/Fix-backward-compatibility-for-klibs-for-ui-backhandler) before beta01, creating a stub project for `backhandler`

MatkovIvan replied:
> creating a stub project for backhandler

the project is still here and contains valid code. What's required is only restoring this dependency

terrakok replied:
```

* What went wrong:
Circular dependency between the following tasks:
:navigationevent:navigationevent-compose:compileDesktopMainJava
+--- :navigationevent:navigationevent-compose:compileKotlinDesktop
|    \--- :compose:ui:ui:desktopJar
|         +--- :compose:ui:ui:compileDesktopMainJava
|         |    +--- :compose:ui:ui:compileKotlinDesktop
|         |    |    \--- :compose:ui:ui-backhandler:desktopJar
|         |    |         +--- :compose:ui:ui-backhandler:compileDesktopMainJava
|         |    |         |    +--- :navigationevent:navigationevent-compose:desktopJar
```

terrakok replied:
I'll try to use a non-compose library then

terrakok replied:
https://github.com/JetBrains/compose-multiplatform-core/pull/2441
